manager/scheduler/nodeset.go

```diff
@@ -45,44 +45,52 @@ func (ns *nodeSet) remove(nodeID string) {
 	delete(ns.nodes, nodeID)
 }
 
-type nodeSorter struct {
-	nodes    []NodeInfo
-	lessFunc func(*NodeInfo, *NodeInfo) bool
-}
-
-func (s nodeSorter) Len() int {
-	return len(s.nodes)
-}
-
-func (s nodeSorter) Swap(i, j int) {
-	s.nodes[i], s.nodes[j] = s.nodes[j], s.nodes[i]
-}
-
-func (s nodeSorter) Less(i, j int) bool {
-	return s.lessFunc(&s.nodes[i], &s.nodes[j])
-}
+func (ns *nodeSet) tree(serviceID string, preferences []*api.PlacementPreference) decisionTree {
+	var root decisionTree
 
-// findBestNodes returns n nodes (or < n if fewer nodes are available) that
-// rank best (lowest) according to the sorting function.
-func (ns *nodeSet) findBestNodes(n int, meetsConstraints func(*NodeInfo) bool, nodeLess func(*NodeInfo, *NodeInfo) bool) []NodeInfo {
-	var nodes []NodeInfo
-	// TODO(aaronl): Is is possible to avoid checking constraints on every
-	// node? Perhaps we should quickselect n*2 nodes that weren't
-	// prescreened, and repeat the quickselect if there weren't enough
-	// nodes meeting the constraints.
 	for _, node := range ns.nodes {
-		if meetsConstraints(&node) {
-			nodes = append(nodes, node)
+		tree := &root
+		for _, pref := range preferences {
+			// Only spread is supported so far
+			spread := pref.GetSpread()
+			if spread == nil {
+				continue
+			}
+
+			descriptor := spread.SpreadDescriptor
+			var value string
+			switch {
+			case len(descriptor) > len(nodeLabelPrefix) && strings.EqualFold(descriptor[:len(nodeLabelPrefix)], nodeLabelPrefix):
+				if node.Spec.Annotations.Labels != nil {
+					value = node.Spec.Annotations.Labels[descriptor[len(nodeLabelPrefix):]]
```

Comment:
> it would cause an enourmous amount of log output unless you make sure that every node has every label that you are selecting on in any service.

I agree. If we take that approach, it should be done outside of the for loop, in a label validation procedure. 

I think keeping empty value making it simpler for now. We can wait to see if it might cause problem.